### PR TITLE
Feature/skill query models

### PR DESCRIPTION
**Pull request by bwolkoys**

### CODE REVIEW

File: server/schemas/typeDefs.js [hunk 1]
```diff
@@ -18,18 +18,29 @@ const typeDefs = `
...
+    skills: [Skill]
+    skill(id: ID!): Skill
+    getSkillRelationships: [SkillRelationship]
   }
 
   type Mutation {
     addUser(username: String!, email: String!, password: String!): Auth
     login(email: String!, password: String!): Auth
+    addSkill(name: String!, description: String!): Skill
+    modifySkill(id: ID!, name: String!, description: String!): Skill
```
Comment (jvalliswalker) [inline server/schemas/typeDefs.js:39]: Can you make the "name" and "description" not required here? Otherwise if I want to modify a skill's name but not its description, I still have to enter in the same description it had before since description is required. The reverse is true with name.

File: server/schemas/typeDefs.js [hunk 2]
```diff
@@ -18,18 +18,29 @@ const typeDefs = `
...
     me: User
+    skills: [Skill]
+    skill(id: ID!): Skill
+    getSkillRelationships: [SkillRelationship]
   }
 
   type Mutation {
     addUser(username: String!, email: String!, password: String!): Auth
     login(email: String!, password: String!): Auth
+    addSkill(name: String!, description: String!): Skill
```
Comment (jvalliswalker) [inline server/schemas/typeDefs.js:38]: Can you make "description" not required here?

File: server/schemas/resolvers.js [hunk 3]
```diff
@@ -13,6 +13,20 @@ const resolvers = {
...
+      //this will grab all the skills from the database
+      return await Skill.find();
+    },
+    skill: async (parent, args, {id}) => {
+      //this will grab a skill by it's id
+      return await Skill.findById(id);
+    },
+    getSkillRelationships: async (parent, args, context) => {
+      //populate taken from module 21, activity 5 /schemas/resolvers.js
+      return await SkillRelationship.find({}).populate('skill').populate('user');
```
Comment (jvalliswalker) [inline server/schemas/resolvers.js:26]: Here you have a query that will return all skill relationships, but in practice we'll be wanting to pull skill relationships by the user they belong to. Otherwise a user trying to view their own skill list (via their skill relationships) will see the skill relationships for every user in the database. Let me know if that makes sense.

Please update to run a "find" based on passed User Id.

File: server/schemas/resolvers.js [hunk 4]
```diff
@@ -13,6 +13,20 @@ const resolvers = {
...
     },
+    skills: async (parent, args, context) => {
+      //this will grab all the skills from the database
+      return await Skill.find();
+    },
+    skill: async (parent, args, {id}) => {
+      //this will grab a skill by it's id
+      return await Skill.findById(id);
+    },
+    getSkillRelationships: async (parent, args, context) => {
```
Comment (jvalliswalker) [inline server/schemas/resolvers.js:24]: I tried running this in the graphql sandbox and got a server error stating "Cannot read properties of undefined (reading 'find')". I did some sleuthing and it looks like the SkillRelationship model is exported in the `Skill_Relationship.js` file, but not imported and re-exported in the `index.js` file in the same folder. Because of this the import of SkillRelationship on line 1 of this file didn't resolve, causing this error. 

You'll need to make that import/export in the `index.js` file to resolve this.

File: server/schemas/typeDefs.js [hunk 5]
```diff
@@ -18,18 +18,29 @@ const typeDefs = `
   }
 
   type SkillRelationship {
-  
+    skill: Skill
+    yearsOfExperience: Int
```
Comment (jvalliswalker) [inline server/schemas/typeDefs.js:22]: So I spent like an hour trying to get decimals to work, but it turns out in graphql/mongodb it's a real pain. Instead of Int, can you change this (and all corresponding schema entries) to String? We can store a decimal as a string and just convert it from Decimal to String on the front end when a user submits data.

File: server/schemas/typeDefs.js [hunk 6]
```diff
@@ -18,18 +18,29 @@ const typeDefs = `
...
+    getSkillRelationships: [SkillRelationship]
   }
 
   type Mutation {
     addUser(username: String!, email: String!, password: String!): Auth
     login(email: String!, password: String!): Auth
+    addSkill(name: String!, description: String!): Skill
+    modifySkill(id: ID!, name: String!, description: String!): Skill
+    deleteSkill(id: ID!): Skill
+    addSkillRelationship(skill: ID!, yearsOfExperience: Int!, areasOfExpertise: String!): SkillRelationship
```
Comment (jvalliswalker) [inline server/schemas/typeDefs.js:41]: For this mutation, we need to have a User Id as a required parameter. Otherwise we  end up with a Skill Relationship record that relates to a skill but not a user. 

Two other small things: 
1. Can you make areasOfExpertise not required?
2. See earlier comment for changing "yearsOfExperience" to a string

File: server/schemas/typeDefs.js [hunk 7]
```diff
@@ -18,18 +18,29 @@ const typeDefs = `
...
+    areasOfExpertise: String
+    user: User
   }
 
   type Query {
     user(id: ID!): User
     me: User
+    skills: [Skill]
+    skill(id: ID!): Skill
+    getSkillRelationships: [SkillRelationship]
```
Comment (jvalliswalker) [inline server/schemas/typeDefs.js:32]: Mentioned in resolvers too, but please add user Id as a required parameter for getSkillRelationships

**Review by jvalliswalker — CHANGES_REQUESTED** [review]
Hi @bwolkoys. Thanks for your work here! I have a number of requested changes in the form of comments. Some of them are minor changes, and some of them are for catching errors/bugs I ran into during testing. 

One thing I wanted to reiterate is the importance of testing your code before moving on to a PR. There are a number of ways to do this, and we can go over some together if you'd like. However it gets done, you need to make sure the code works as desired before handing it off. This means spinning up the server and walking through each use case for the code your writing. With this pull specifically, you need to test each query and mutation after writing it, and check for edge cases. I did this for this PR via the graphql sandbox endpoint since we don't have a UI for testing these queries yet.

File: server/schemas/typeDefs.js (hunk 1 above)
Comment (bwolkoys) [reply server/schemas/typeDefs.js:39]: this comment had been updated in feature/profile-page

File: server/schemas/typeDefs.js (hunk 2 above)
Comment (bwolkoys) [reply server/schemas/typeDefs.js:38]: this comment had been updated in feature/profile-page

File: server/schemas/resolvers.js (hunk 3 above)
Comment (bwolkoys) [reply server/schemas/resolvers.js:26]: this comment had been updated in feature/profile-page

File: server/schemas/resolvers.js (hunk 4 above)
Comment (bwolkoys) [reply server/schemas/resolvers.js:24]: this comment had been updated in feature/profile-page

File: server/schemas/typeDefs.js (hunk 5 above)
Comment (bwolkoys) [reply server/schemas/typeDefs.js:22]: this comment had been updated in feature/profile-page

File: server/schemas/typeDefs.js (hunk 6 above)
Comment (bwolkoys) [reply server/schemas/typeDefs.js:41]: this comment had been updated in feature/profile-page

File: server/schemas/typeDefs.js (hunk 7 above)
Comment (bwolkoys) [reply server/schemas/typeDefs.js:32]: this comment had been updated in feature/profile-page

**Comment by jvalliswalker** [comment]
Closed per transition to PR https://github.com/UCBX-2024-Team-Wombat/Final-Project/pull/12
